crl/validator.go

```diff
@@ -20,346 +20,420 @@ package crl
-	}
-
-	crl, err := x509.ParseRevocationList(data)
-	if err != nil {
-		return fmt.Errorf("unable to parse downloaded CRL (url=%s): %w", endpoint, err)
+	parseCertificatesSlice := func(rawCerts [][]byte) ([]*x509.Certificate, error) {
+		var raw []byte
+		for _, rawCert := range rawCerts {
+			raw = append(raw, rawCert...)
+		}
+		return x509.ParseCertificates(raw)
 	}
 
-	if err := v.verifyCRL(crl); err != nil {
-		return fmt.Errorf("CRL verification failed (issuer=%s): %w", crl.Issuer.String(), err)
+	// check that all cert issuers are in the truststore, and add missing crl distribution points.
+	for _, chain := range config.Certificates {
+		certificates, err := parseCertificatesSlice(chain.Certificate)
+		if err != nil {
+			return err
+		}
+		for _, cert := range certificates {
+			issuer := v.truststore[cert.Issuer.String()]
+			if issuer == nil {
+				// This indicates a mismatch between crlValidator truststore and tls.Config. This is a programming error.
+				return errors.New("tls.Config contains certificate from issuer that is not in the truststore: " + cert.Subject.String())
+			}
+			// add any previously unknown CRL distribution points.
+			// (Leaf cert is likely to contain CRLDistributionPoints not found in the truststore.)
+			if len(cert.CRLDistributionPoints) > 0 {
+				for _, endpoint := range cert.CRLDistributionPoints {
+					if v.crls[endpoint] == nil {
+						v.crls[endpoint] = newRevocationList(issuer)
+					}
+				}
+			}
+		}
 	}
 
-	v.listsLock.Lock()
-	defer v.listsLock.Unlock()
+	config.VerifyPeerCertificate = func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		// TODO: the truststore should already contain ALL CAs that might be found in the verifiedChains
+		// TODO: should we check rawCerts, or verified chains?
+		// rawCerts contains raw cert data from the tls handshake.
+		// It is used together with the tls.Config to generate verifiedChains, but may contain additional certs that are not in a verified chain.
 
-	issuerName := crl.Issuer.String()
+		certificates, err := parseCertificatesSlice(rawCerts)
+		if err != nil {
+			return err
+		}
 
-	for _, cert := range crl.RevokedCertificates {
-		v.setRevoked(issuerName, cert.SerialNumber)
+		return v.validateChain(certificates)
 	}
-
-	v.lists[endpoint] = crl
-
 	return nil
 }
 
-// crlHasExpired is a replacement for pkix.CertificateList.HasExpired
-func crlHasExpired(crl *x509.RevocationList, deadline time.Time) bool {
-	return !deadline.Before(crl.NextUpdate)
+// validateChain validates that none of the certificates in the chain is revoked.
+// Certificated in chain are assumed ordered from leaf to root certificate.
+func (v *validator) validateChain(chain []*x509.Certificate) error {
+	var cert *x509.Certificate
+	var err error
+	for i := range chain {
+		cert = chain[len(chain)-1-i]
+		// check in reverse order to prevent CRL expiration errors due to revoked CAs no longer issuing CRLs
```

Comment:
is the order consistent?

Reply:
tls handshake provides them sorted from leaf to CA. Best effort approach imo